src/python/pants/backend/python/providers/pbs/rules.py

```diff
@@ -0,0 +1,237 @@
+        """
+    )
+
+    known_python_versions = StrListOption(
+        default=None,
+        default_help_repr=f"<Metadata for versions: {', '.join(sorted(load_pbs_pythons()))}>",
+        advanced=True,
+        help=textwrap.dedent(
+            f"""
+            Known versions to verify downloads against.
+
+            Each element is a pipe-separated string of `version|platform|sha256|length|url`, where:
+
+            - `version` is the version string
+            - `platform` is one of `[{','.join(Platform.__members__.keys())}]`
+            - `sha256` is the 64-character hex representation of the expected sha256
+                digest of the download file, as emitted by `shasum -a 256`
+            - `length` is the expected length of the download file in bytes, as emitted by
+                `wc -c`
+            - `url` is the download URL to the `.tar.gz` archive
+
+            E.g., `3.1.2|macos_x86_64|6d0f18cd84b918c7b3edd0203e75569e0c7caecb1367bbbe409b44e28514f5be|42813|https://<URL>`.
+
+            Values are space-stripped, so pipes can be indented for readability if necessary.
+
+            Additionally, any versions you specify here will override the default Pants metadata for
+            that version.
+            """
+        ),
+    )
+
+    def get_all_pbs_pythons(self) -> dict[str, dict[str, PBSPythonInfo]]:
+        all_pythons = load_pbs_pythons().copy()
+
+        for version_info in (self.known_python_versions or []):
+            try:
+                pyversion, platform, sha256, filesize, url = [
+                    x.strip() for x in version_info.split("|")
+                ]
+            except ValueError:
+                raise ExternalToolError(
+                    f"Bad value for [{PBSPythonProviderSubsystem.options_scope}].known_python_versions: {version_info}"
+                )
+
+            if pyversion not in all_pythons:
+                all_pythons[pyversion] = {}
+
+            all_pythons[pyversion][platform] = PBSPythonInfo(
+                url=url, sha256=sha256, size=int(filesize)
+            )
+
+        return all_pythons
+
+
+class PBSPythonProvider(PythonProvider):
+    pass
+
+
+def _choose_python(
+    interpreter_constraints: InterpreterConstraints,
+    universe: Iterable[str],
+    pbs_versions: collections.abc.Collection[str],
+) -> str:
+    """Choose the highest supported patch of the lowest supported Major/Minor version."""
+    supported_python_triplets = interpreter_constraints.enumerate_python_versions(universe)
+    version_triplet: tuple[int, int, int] | None = None
+    for triplet in supported_python_triplets:
+        pbs_supported_version = ".".join(map(str, triplet)) in pbs_versions
+        if pbs_supported_version:
+            if version_triplet and version_triplet[:2] < triplet[:2]:
+                # This version is a major/minor above the previous supported one, we're done.
+                break
+
+            version_triplet = triplet
```

Comment:
This loops seems to rely on `enumerate_python_versions` returning versions in order; is this the case?

Reply:
It is, and I expect that's part of it's contract.